Make earliest ledger sequence configurable

miguelportilla opened the pull request.

mellery451 commented on src/ripple/nodestore/impl/Database.cpp:
```diff
@@ -25,12 +25,27 @@
+    std::uint32_t gs;
+    if (get_if_exists<std::uint32_t>(config, "genesis_seq", gs))
+    {
+        if (genesisSeq_ != 32570)
```
minor: maybe define a constexpr value for this default genesisSeq value?

bachase commented on doc/rippled-example.cfg:
```diff
@@ -824,6 +824,10 @@
 #
+#       genesis_seq         The earliest ledger sequence to start from. The
+#                           ripple ledger consensus network default is
+#                           32570, alternate networks may set this value.
```
Consider rephrasing to make it clear is 32570 is the application default. As written, the reader might believe 32570 is only the default when connecting to the main network. Perhaps:
```
The default is 32570 to match the XRP ledger network's earliest allowed sequence. Alternate networks may set this value.
```

miguelportilla replied:
Thanks. 👍

bachase commented on src/ripple/nodestore/impl/Database.cpp:
```diff
@@ -25,12 +25,27 @@
+    if (get_if_exists<std::uint32_t>(config, "genesis_seq", gs))
+    {
+        if (genesisSeq_ != 32570)
+            Throw<std::runtime_error>("genesis_seq set more than once");
```
How would a user set the genesis sequence more than once? Just trying to understand exactly what this checks against.

miguelportilla replied:
With the Node store and the shard store.

bachase commented on src/ripple/nodestore/Database.h:
```diff
@@ -209,11 +209,24 @@ class Database : public Stoppable
 
+    // The earliest ledger sequence. Default is 32570
+    // May be set otherwise for tests or alternate network
+    static std::uint32_t genesisSeq_;
```
Is there any thread safety issues? I don't see any and suspect you've already thought about it, but want to raise anyway.

miguelportilla replied:
No, but I'll make the member private for additional safety.

bachase commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -93,8 +93,12 @@ DatabaseShardImp::init()
+        if (shardIndex < genesisShardIndex_)
+        {
+            JLOG(j_.fatal()) <<
+                "Invalid shard index";
```
Maybe add the `shardIndex` and `genesisShardIndex` to the error message?

miguelportilla commented:
@bachase, @mellery451 Thanks for the feedback. I pushed a commit addressing comments. I'll work on extending the database unit test to add coverage for this new feature.

miguelportilla commented:
@bachase @mellery451 Unit test added.

bachase commented:
I'm getting some unit test failures
```
ripple.NodeStore.Database NodeStore backend 'memory'
#4 failed: Database_test.cpp(178)
ripple.NodeStore.Database NodeStore backend 'nudb'
#1 failed: Database_test.cpp(121)
#6 failed: unhandled exception: genesis_seq set more than once
285ms, 1 suite, 2 cases, 12 tests total, 3 failures
```

mellery451 commented on src/test/nodestore/Database_test.cpp:
```diff
@@ -156,32 +159,52 @@ class Database_test : public TestBase
-        testNodeStore ("rocksdb", true, seedValue);
-    #endif
-    }
+        {
```
nit - it don't think this outer block serves any purpose.

miguelportilla replied:
👍 removed

miguelportilla commented:
Fixed test faile issue.

mellery451 commented:
tests look good now, getting one warning that looks easy to fix:

```
src/test/nodestore/Database_test.cpp|207 col 22 warning| unused variable 'i' [-Wunused-variable]
                 auto i = std::strcmp(e.what(),
```

miguelportilla commented:
@mellery451 Yikes, that line with the warning ws left over debug code. Thanks.

bachase commented:
👍

ripplelabs-jenkins commented:
## Jenkins Build Summary

Built from [this commit](https://github.com/ripple/rippled/commit/143cc8f9f11d508d2f64507a9a1608c4722aacbe)

Built at __20180301 - 06:41:40__

### Test Results

Build Type | Result | Status
---------- | ------ | ------

MarkusTeufelberger commented on doc/rippled-example.cfg:
```diff
@@ -824,6 +824,10 @@
 #
+#       genesis_seq         The default is 32570 to match the XRP ledger
+#                           network's earliest allowed sequence. Alternate
+#                           networks may set this value. Minimum value of 1. 
```
Isn't the genesis ledger usually ledger 0? Not that it makes much of a difference, especially in sharding, but it seems weird to start at height 1.

miguelportilla commented:
@nbougalis Thanks for the feedback. I pushed a commit that addresses the issues.

@nbougalis @bachase @mellery451 I've included a commit from David that improves shard sharing with peers more than a hop away. It also fixes an old bug where we don't check to make sure we don't route a ledger request back to the peer we got it from. I have reviewed his changes and given them the green light.

bachase commented on src/ripple/nodestore/DatabaseShard.h:
```diff
@@ -45,12 +45,17 @@ class DatabaseShard : public Database
+        : Database(name, parent, scheduler, readThreads, config, journal)
     {
         get_if_exists<std::uint32_t>(config, "ledgers_per_shard", lps_);
+        genesisShardIndex_ = seqToShardIndex(earliestSeq());
```
Does this (and the line above) share @nbougalis's concern of initializing a static member each time this constructor is called? Even if its guaranteed to only be called once now, that seems a fragile/hard to enforce requirement.

miguelportilla replied:
The issue is a side effect of making the seq configurable. It definitely introduced some ugliness, I'll fix it.

bachase commented on src/ripple/nodestore/DatabaseShard.h:
```diff
@@ -159,16 +187,21 @@ class DatabaseShard : public Database
     static std::uint32_t lps_;
+
+    // The earliest shard index
+    static std::uint32_t genesisShardIndex_;
```
Nit: Rename `genesisShardIndex_` to `earliestShardIndex_`

bachase commented on src/ripple/overlay/impl/PeerImp.cpp:
```diff
@@ -2172,40 +2172,40 @@ PeerImp::getLedger (std::shared_ptr<protocol::TMGetLedger> const& m)
+                // Search for a peer who might
+                auto const v = getPeerWithLedger(overlay_, ledgerhash,
+                    packet.has_ledgerseq() ? packet.ledgerseq() : 0, this);
+                if (! v || (v.get() == this))
```
`v.get() == this` means we returned ourselves as the peer? Does that warrant its own `JLOG` message?

miguelportilla replied:
It is not an error and logging it would cause superfluous messaging. It might be worth extending getPeerWithLedger to optionally exclude `this` but that beyond this PR.

bachase replied:
But aren't we logging now on line 2200?

bachase replied:
More directly, I was wondering if the existing logging message should distinguish the case of `v` being null or `v` being ourselves.

miguelportilla replied:
Sorry, I misunderstood you. I am glad you brought this bit to my attention. Upon closer inspection, it is obvious the code is unnecessary and the previous enhancement I mentioned is already in place! How did we miss this? Some Self-flagellation in order for the lot!

eg.
`if (! v || (v.get() == this))` should stay as is `if (!v)`

bachase commented:
Not sure if this helps, but to me, the  `genesis` ledger is the ledger that has no parent and is the initial/first ledger. Then `earliest` ledger is the ledger closest to the genesis that we believe is available.

ximinez commented on src/ripple/nodestore/impl/Shard.h:
```diff
@@ -46,9 +46,8 @@ using NCache = KeyCache<uint256>;
-    Shard(std::uint32_t index, int cacheSz,
-        PCache::clock_type::rep cacheAge,
-        beast::Journal& j);
+    Shard(DatabaseShard const& db, std::uint32_t index, int cacheSz,
```
On non-unity builds, `DatabaseShard` is not defined here. A pre-declaration around line 40 appears to resolve the issue.

miguelportilla replied:
@ximinez Thanks, pushed fix.

ximinez commented:
I've added my approval, but I didn't review the whole PR. This just acknowledges that my issue was resolved.

bachase commented:
In 1.0.0-b3 as 0b18b36186797152f49d0dd4e388b8ad98c04820
 and 5db5e3114014173113324ff885a5859053fd0eb4
